glib/src/main_context_futures.rs

```diff
@@ -230,19 +246,199 @@ impl TaskSource {
 
                 // This will panic if the future was a local future and is called from
                 // a different thread than where it was created.
-                Pin::new(&mut self.future).poll(&mut context)
+                if let Some(tx) = self.return_tx.take() {
+                    let res = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+                        Pin::new(&mut self.future).poll(&mut context)
+                    }));
+                    match res {
+                        Ok(Poll::Ready(res)) => {
+                            let _ = tx.send(Ok(res));
+                            Poll::Ready(())
+                        }
+                        Ok(Poll::Pending) => {
+                            self.return_tx.replace(tx);
+                            Poll::Pending
+                        }
+                        Err(e) => {
+                            let _ = tx.send(Err(e));
+                            Poll::Ready(())
+                        }
+                    }
+                } else {
+                    Pin::new(&mut self.future).poll(&mut context).map(|_| ())
+                }
             })
             .expect("current thread is not owner of the main context")
     }
 }
 
+// rustdoc-stripper-ignore-next
+/// A handle to a task running on a [`MainContext`].
+///
+/// Like [`std::thread::JoinHandle`] for a task rather than a thread. The return value from the
+/// task can be retrieved by awaiting on this object. Dropping the handle "detaches" the task,
+/// allowing it to complete but discarding the return value.
+#[derive(Debug)]
+pub struct JoinHandle<T> {
+    rx: oneshot::Receiver<std::thread::Result<Box<dyn Any + 'static>>>,
+    source: Source,
+    id: Cell<Option<NonZeroU32>>,
+    phantom: PhantomData<oneshot::Receiver<std::thread::Result<T>>>,
+}
+
+impl<T> JoinHandle<T> {
+    #[inline]
+    fn new(
+        ctx: &MainContext,
+        source: Source,
+        rx: oneshot::Receiver<std::thread::Result<Box<dyn Any + 'static>>>,
+    ) -> Self {
+        let id = source.attach(Some(ctx));
+        let id = Cell::new(Some(unsafe { NonZeroU32::new_unchecked(id.as_raw()) }));
+        Self {
+            rx,
+            source,
+            id,
+            phantom: PhantomData,
+        }
+    }
+    // rustdoc-stripper-ignore-next
+    /// Returns the internal source ID.
+    ///
+    /// Returns `None` if the handle was aborted already.
+    #[inline]
+    pub fn as_raw_source_id(&self) -> Option<u32> {
+        self.id.get().map(|i| i.get())
+    }
+    // rustdoc-stripper-ignore-next
+    /// Aborts the task associated with the handle.
+    #[inline]
+    pub fn abort(&self) {
+        if let Some(id) = self.id.take() {
+            let id = unsafe { SourceId::from_glib(id.get()) };
+            id.remove();
```

Comment:
This is not correct if a non-default main context is used (and also is not thread-safe). What you want here is `self.source.destroy()` instead

Reply:
Changed it, but does that mean `Source` should be `!Send`?

Reply:
Sorry not `Source`, I meant `SourceId`?

Reply:
Sending it to other threads is only one of the many ways how you can make it fail. Ideally the API should just not exist at all :)

Reply:
Ok, let's save that for later https://github.com/gtk-rs/gtk-rs-core/issues/860